---
title: "WAWBE_core_dataset_creation"
author: "Erin Dahl"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

print(Sys.time())


```

## Purpose

The goal of this code is to calculate WW trends. 

## Libraries

```{r libraries, message=FALSE}

library(tidyverse)
library(REDCapR)
library(slider)
library(here)
library(sf)

time_start <- now()

```


## Load paths and data

```{r load_data}

# file containing redcap credentials, see credentials_ref.R for the variables needed
source(here("credentials.R"))

# source calculation functions from nwss
source(here("wwstrendfunctions.R"))

# load the population df for the sewersheds, please contact our team for your file
load(here("population.RData")) 

```

## Read and merge redacap data

Read in redcap data and explore current redcap variables read in to R

```{r redcap}

# retrieve the ww samples redcap data
redcap_ww_samples <- redcap_read(redcap_uri=redcap_url, token=ww_samples_token)$data

# ww sample data from non-repeating instruments
redcap_samples<- redcap_ww_samples %>% 
                      filter(is.na(redcap_repeat_instrument)) %>%
                      # remove columns
                      select(-c(redcap_repeat_instrument,
                                test_result_date,
                                redcap_repeat_instance,
                                starts_with("pcr"),
                                lod_sewage,
                                quality_flag,
                                contains(c("variant", "complete", "timestamp", 
                                           "received", "printed",
                                           "orig")))) 

# pcr target associated data from c3_pcr_target_report_form
redcap_targets <- redcap_ww_samples %>% 
                      filter(redcap_repeat_instrument =="c3_pcr_target_report_form") %>%
                      # keep only desired columns
                      select(sample_id, 
                             test_result_date,
                             starts_with("pcr"),
                             lod_sewage,
                             quality_flag) %>%
                      # map values
                      mutate(pcr_target = case_match(pcr_target, 
                                                      1 ~	"sars-cov-2",
                                                      2	~ "delta",
                                                      3	~ "omicron",
                                                      4	~ "hMPXV",
                                                      5	~ "FLUAV",
                                                      6	~ "FLUBV",
                                                      7 ~ "RSV"),
                             pcr_gene_target = case_match(pcr_gene_target,
                                                          1	~ "n1",
                                                          2	~ "n2",
                                                          3	~ "p681r",
                                                          4	~ "del156-157",
                                                          5	~ "wt214",
                                                          6	~ "del69/70",
                                                          7	~ "ins214epe",
                                                          8	~ "del142-144",
                                                          9	~ "wt156-157",
                                                          10 ~ "e9l-nvar",
                                                          11 ~ "G2R_G",
                                                          12 ~ "InfA1",
                                                          13 ~ "InfA2",
                                                          14 ~ "InfB",
                                                          15 ~ "n",
                                                          16 ~ "InfA1 and InfA2 combined",
                                                          17 ~ "RSV-A and RSV-B combined"))


# merge the target specific and sample specific variables
redcap_all_samples <- redcap_targets %>% 
                          merge(redcap_samples, by = "sample_id") %>%
                          mutate(sample_id = as.character(sample_id)) %>%
                          # rename the sample site name to match naming in the population df
                          rename(site_id = sample_site_name) %>% 
                          # join to add the population data
                          left_join(population, by = "site_id")  %>% 
                          # format the date
                          mutate(sample_collect_date = as.Date(sample_collect_date)) %>% 
                          # remove data missing sample_site_ids
                          filter(!is.na(site_id)) %>%
                          # group by site id and gene target
                          group_by(site_id,pcr_gene_target, .drop = F) %>% 
                          # keep only first ww record if there are two for a site on the same day
                          distinct(sample_collect_date, .keep_all = TRUE) %>%
                          # targets tested, with no positive detection will
                          # have a missing conc value. To show no detection, change NA to 0
                          mutate(pcr_target_avg_conc = if_else(is.na(pcr_target_avg_conc),0, pcr_target_avg_conc))

```


## Calculate wastewater trends

 - normalized flow
 - log normalized flow
 - determine trend categories

```{r ww_trends}

# map the population data to calculate trends
wwdata_flow_calc <- redcap_all_samples %>%  
                          # calculate flow rate in Liters
                          # original flow rate is in Million gallons per day; 
                          # 1 million gallon/day [US] to liters per day = 3785411.8 liters per day updated 2022-09-12
                          # If the pcr_target_avg_conc is less than the lod_sewage, 
                          # then the adjusted concentration value is 1/2 the lod sewage value
                          # This value change is is alignment with how CDC NWSS handles samples with concentrations below lod
                            mutate(flow_rate_l = flow_rate*3785411.8, 
                                   pcr_target_avg_conc_adj = if_else(pcr_target_avg_conc < lod_sewage, 0.5 * lod_sewage, pcr_target_avg_conc),
                                   pcr_target_norm_conc = (pcr_target_avg_conc_adj*flow_rate_l)/totalpop, 
                                   pcr_target_log_norm_conc = log10(pcr_target_norm_conc)) %>%  
                            # filter data with log normalized conc > 0
                            filter(pcr_target_log_norm_conc > 0) 

# Calculate the adjustment values for special handling of reported 0 concentration
adjustment_values <- wwdata_flow_calc %>% 
  filter(pcr_target_avg_conc > 0) %>% 
  group_by(pcr_gene_target) %>% 
  summarise(mean= mean(pcr_target_log_norm_conc), 
            min = min(pcr_target_log_norm_conc), 
            max = max(pcr_target_log_norm_conc), 
            adj_log_norm_conc = 0.75 * min(pcr_target_log_norm_conc))   

# Replace pcr_target_log_norm_conc with the calculated adjusted value when the
# reported concentration is 0
wwdata_flow_calc_adjusted <- wwdata_flow_calc %>%
  left_join(adjustment_values, by = "pcr_gene_target") %>%
  mutate(
    pcr_target_log_norm_conc = if_else(pcr_target_avg_conc == 0, 
                                       adj_log_norm_conc, 
                                       pcr_target_log_norm_conc)) %>%
  select(-adj_log_norm_conc)

# calculate trends for each sample site id
# if desired, only calculate trends for a specific gene target
wwtrends_cat <-  wwdata_flow_calc_adjusted %>% #filter(pcr_gene_target == "n1") %>% 
                                  group_by(site_id, pcr_gene_target) %>% 
                                  # calculate trends for each site/gene target combination
                                  group_modify(~ calc_classify_trends(.data = .,
                                       .date = "sample_collect_date",
                                       .trend_var ='pcr_target_log_norm_conc' ,
                                       .window_type = "time",
                                       .window_size_sus = 15, 
                                       .window_size_short = 8)) %>%
                                  # using the sustained window (15) to categorize trends into 7 categories
                                  # NA data in percent_cat means there is no recent data
                                  mutate(ptc_sus = round(ptc_sus, 0),
                                         percent_cat = cut(ptc_sus, 
                                                           breaks = c(-1000000000, -100, -10, 0, 9, 99, 999, Inf), 
                                                           labels = c("-100%", 
                                                                      "-99% to -10%", 
                                                                      "-9% to 0%", 
                                                                      "1% to 9%", 
                                                                      "10% to 99%", 
                                                                      "100% to 999%", 
                                                                      "1000% or more"),
                                                           include.lowest = T,
                                                           right = T)) %>%
                                select(-ends_with(c("sus", "short")))

```

## Create finalized data object

```{r core_data}

keep_flow_calc <- wwdata_flow_calc %>% select(site_id, 
                                                  sample_collect_date,
                                                  test_result_date,
                                                  import_redcap_date,
                                                  micro_lab_name,
                                                  pcr_target,
                                                  pcr_gene_target,
                                                  pcr_target_avg_conc,
                                                  pcr_target_norm_conc,
                                                  pcr_target_below_lod,
                                                  flow_rate_l,
                                                  totalpop,
                                                  viral_data_flag,
                                                  data_flag_notes) 


# add ww metrics of interest back in the dataset and location data
core_dataset <-  wwtrends_cat %>% rename(sample_collect_date = date) %>% 
                                          left_join(x=.,
                                                    y= keep_flow_calc,
                                                    by= c("sample_collect_date", "site_id", "pcr_gene_target")) %>%
                                          ungroup() 



```

## Save core_dataset

```{r save}
# The line below saves an R object, alternatively the df could be saved as a csv
# save(core_dataset, file = core_dataset_path2)

time_end <- now()

# total time to run script
time_end - time_start
```
